R/type-bare.R

```diff
@@ -86,8 +86,12 @@ vec_ptype2.list <- function(x, y, ..., x_arg = "x", y_arg = "y") {
 #' @method vec_ptype2.logical logical
 #' @export
 vec_ptype2.logical.logical <- function(x, y, ..., x_arg = "x", y_arg = "y") {
+  # Special case `vec_ptype2(NA, NA)` to ensure that
+  # `unspecified()` is returned
   if (is.object(y)) {
     vec_default_ptype2(x, y, x_arg = x_arg, y_arg = y_arg)
+  } else if (is_unspecified(x) && is_unspecified(y)) {
```

Comment:
Hmm, `shape_match()` returns `type` immediately rather than shape matching if either `x` or `y` are objects, which in this case they both are since unspecified vectors are classed.

I could unclass them?

```r
shape_match(unspecified(), unclass(x), unclass(y))
```

Reply:
> Should a logical matrix of all NA values be considered unspecified?

I think so. If it's too much work, we could deal with it another time.

Reply:
It breaks a few tests (in particular, one in `vec_rbind()` that looks like it could use some thought), I'm going to leave it for a separate pr